akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -132,6 +134,152 @@ import scala.util.control.NonFatal
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      var accumulated = List.empty[In]
+      var subSource: SubSourceOutlet[In] = null
+      var subSink: SubSinkInlet[Out] = null
+
+      setHandlers(in, out, this)
+
+      override def postStop(): Unit = {
+        matPromise.tryFailure(new AbruptStageTerminationException(this))
+        if (null != subSource && !subSource.isClosed)
+          subSource.complete()
+        if (null != subSink && !subSink.isClosed)
+          subSink.cancel()
+        super.postStop()
+      }
+
+      override def onPush(): Unit = {
+        if (null != subSource) {
+          subSource.push(grab(in))
+        } else {
+          accumulated ::= grab(in)
+          if (accumulated.size == n) {
+            materializeFlow()
+          } else {
+            //gi'me some more!
+            pull(in)
+          }
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (null != subSource) {
+          subSource.complete()
+        } else {
+          materializeFlow()
+        }
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        if (null != subSource) {
+          subSource.fail(ex)
+        } else {
+          //flow won't be materialized, so we have to complete the future with a failure indicating this
+          matPromise.failure(new NeverMaterializedException(ex))
+          super.onUpstreamFailure(ex)
+        }
+      }
+
+      override def onPull(): Unit = {
+        if (null != subSink) {
+          //delegate to subSink
+          subSink.pull()
+        } else if (accumulated.size < n) {
+          pull(in)
+        } else if (accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
+          materializeFlow()
+        }
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        if (null != subSink) {
+          subSink.cancel(cause)
+        } else {
+          materializeFlow()
+        }
+      }
+
+      def materializeFlow(): Unit = {
+        val prefix = accumulated.reverse
+        accumulated = Nil
+        subSource = new SubSourceOutlet[In]("subSource")
+        subSource.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = {
+              if (!isClosed(in) && !hasBeenPulled(in)) {
+                pull(in)
+              }
+            }
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+              //super.onDownstreamFinish(cause)
```

Comment:
notice this is the sub-source's outlet, its downstream is the materialized flow, so this downstream-finished signal comes from the materialized flow - in case in is still open we propagate the signal upstream, if it's already closed we can't (I think I've introduced this condition after experiencing some test failures).

I'll revisit the existing tests and c if your suggested use-case isn't already covered.

Reply:
I may have mixed it up. 

So if a cancellation comes from the inner flow, and upstream is completed already, the cancellation could either come from the inner flow itself, `fromSinkAndSource` for example, which has its downstream end still running, and in that case we should ignore the cancellation here (could be that we should act different if it is a propagated failure from downstream rather than just the normal cancel which will always be of type `akka.stream.SubscriptionWithCancelException.NonFailureCancellation`) and let the inner flow continue running, but it could also come from the actual downstream and have passed up through the inner flow, in that case out will be closed so we can check that as well.

I may be missing some additional complexity of this though.

Reply:
I don't think we should care too much about this:
1. if in is still open, we propagate the signal upstream
2. if in is already closed, the materialized flow has already processed this signal and probably decided to keep going and completed just now.
3. if out is already closed, then closing in indicates all stages have completed, hence the stage can be closed. next hing that happens is a call to onStop. M I missing anything?

can u think of a flow that closes both ends but keeps running? if such a case exists we might see the outer stage being closed (postStop called) while the inner stage is still running, I'm not sure what this is going to do...

Reply:
Isn't the problem that since you swallow the cancellation here if in is closed, so the stage will never actually stop but keep running, or what triggers stopping it?

I think the problematic scenario would be:
1. stage consumes prefix, materializes inner flow
2. source completes
3. downstream cancels or fails
4. cancel or fail comes through inner flow to here
5. in is closed, inner flow is probably stopped but this stage keeps running (leak)

You should be able to detect such a scenario with probes and having the test complete, assertAllStagesStopped wrapping the test should notice the graph is still alive and fail.

Reply:
as demonstrated in test case 'run a detached flow',
when both in and out are closed the 'external' stage stops, in an extreme (made up?) scenario when the materialized flow keeps running, the external stage stops while the nested flow keeps running. when nested flow completes its stages are simply shutdown since inlets and outlets are being closed so active ports drops to zero for these stages and the interpreter simply closes them.